### PR TITLE
smb: adds fingerprint keyword

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
TODO

Describe changes:
- smb: adds fingerprint keyword, Kind of like TLS ja3 or hassh, for SMB

suricata-verify-pr: 844

### CODE REVIEW

File: doc/userguide/rules/smb-keywords.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
+SMB Keywords
+==============
+
+smb.fingerprint
+---------------
+
+Match on a fingerprint of a SMB client, kind of like ssh.hassh or TLS ja3.
```
Comment (jlucovsky) [inline doc/userguide/rules/smb-keywords.rst:7]: Do we want to include the size restriction for the SMB fingerprint?

Do you want to mention that `smb.fingerprint-client` is an alias?

File: rust/src/smb/log.rs [hunk 2]
```diff
@@ -136,6 +139,11 @@ fn smb_common_header(jsb: &mut JsonBuilder, state: &SMBState, tx: &SMBTransactio
 
     match tx.type_data {
         Some(SMBTransactionTypeData::SESSIONSETUP(ref x)) => {
+            let mut outbuf: [u8; 16] = [0; 16];
+            if let Ok(_) = smb_get_fingerprint(&state, tx, &mut outbuf) {
+                jsb.set_string("fingerprint", &to_hex(&outbuf[..]))?;
```
Comment (jlucovsky) [inline rust/src/smb/log.rs:144]: Need to add this to the `eve.json` schema?

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on tlpr1_alerts_cmp.
  

Pipeline 7733

File: doc/userguide/rules/smb-keywords.rst (hunk 1 above)
Comment (catenacyber) [reply doc/userguide/rules/smb-keywords.rst:7]: > Do we want to include the size restriction for the SMB fingerprint?

Which size restriction do you mean ?

> Do you want to mention that smb.fingerprint-client is an alias?

If you want me to ;-)

File: rust/src/smb/log.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/smb/log.rs:144]: Indeed, force-pushed a change to get CI running again

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on tlpr1_alerts_cmp.
  

Pipeline 7751

File: src/detect-smb-fp.c [hunk 3]
```diff
@@ -0,0 +1,382 @@
+/* Copyright (C) 2021 Open Information Security Foundation
```
Comment (regit) [inline src/detect-smb-fp.c:1]: Copyright date is off ;)

File: src/detect-smb-fp.c [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+}
+
+#ifdef UNITTESTS
+void DetectSMBfingerprintRegisterTests(void);
+#endif
+
+void DetectSmbFingerprintRegister(void)
+{
+    sigmatch_table[DETECT_SMB_FINGERPRINT].name = "smb.fingerprint";
+    sigmatch_table[DETECT_SMB_FINGERPRINT].alias = "smb.fingerprint-client";
```
Comment (jasonish) [inline src/detect-smb-fp.c:364]: Why an alias for a brand new keyword?  I typically think of the alias feature for when we do something like changing the style of keyword naming, but for new features we should probably just provide one way to do it.

File: src/detect-smb-fp.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-smb-fp.c:1]: fixing these

**Pull request closed without merging**